executor/slow_query.go

```diff
@@ -34,56 +37,57 @@ import (
-	{execdetails.ResolveLockTimeStr, mysql.TypeDouble, 22, 0, nil, nil},
-	{execdetails.LocalLatchWaitTimeStr, mysql.TypeDouble, 22, 0, nil, nil},
-	{execdetails.WriteKeysStr, mysql.TypeLonglong, 22, 0, nil, nil},
-	{execdetails.WriteSizeStr, mysql.TypeLonglong, 22, 0, nil, nil},
-	{execdetails.PrewriteRegionStr, mysql.TypeLonglong, 22, 0, nil, nil},
-	{execdetails.TxnRetryStr, mysql.TypeLonglong, 22, 0, nil, nil},
-	{execdetails.ProcessTimeStr, mysql.TypeDouble, 22, 0, nil, nil},
-	{execdetails.WaitTimeStr, mysql.TypeDouble, 22, 0, nil, nil},
-	{execdetails.BackoffTimeStr, mysql.TypeDouble, 22, 0, nil, nil},
-	{execdetails.LockKeysTimeStr, mysql.TypeDouble, 22, 0, nil, nil},
-	{execdetails.RequestCountStr, mysql.TypeLonglong, 20, mysql.UnsignedFlag, nil, nil},
-	{execdetails.TotalKeysStr, mysql.TypeLonglong, 20, mysql.UnsignedFlag, nil, nil},
-	{execdetails.ProcessKeysStr, mysql.TypeLonglong, 20, mysql.UnsignedFlag, nil, nil},
-	{variable.SlowLogDBStr, mysql.TypeVarchar, 64, 0, nil, nil},
-	{variable.SlowLogIndexNamesStr, mysql.TypeVarchar, 100, 0, nil, nil},
-	{variable.SlowLogIsInternalStr, mysql.TypeTiny, 1, 0, nil, nil},
-	{variable.SlowLogDigestStr, mysql.TypeVarchar, 64, 0, nil, nil},
-	{variable.SlowLogStatsInfoStr, mysql.TypeVarchar, 512, 0, nil, nil},
-	{variable.SlowLogCopProcAvg, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopProcP90, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopProcMax, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopProcAddr, mysql.TypeVarchar, 64, 0, nil, nil},
-	{variable.SlowLogCopWaitAvg, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopWaitP90, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopWaitMax, mysql.TypeDouble, 22, 0, nil, nil},
-	{variable.SlowLogCopWaitAddr, mysql.TypeVarchar, 64, 0, nil, nil},
-	{variable.SlowLogMemMax, mysql.TypeLonglong, 20, 0, nil, nil},
-	{variable.SlowLogSucc, mysql.TypeTiny, 1, 0, nil, nil},
-	{variable.SlowLogPlan, mysql.TypeLongBlob, types.UnspecifiedLength, 0, nil, nil},
-	{variable.SlowLogPlanDigest, mysql.TypeVarchar, 128, 0, nil, nil},
-	{variable.SlowLogPrevStmt, mysql.TypeLongBlob, types.UnspecifiedLength, 0, nil, nil},
-	{variable.SlowLogQuerySQLStr, mysql.TypeLongBlob, types.UnspecifiedLength, 0, nil, nil},
+//SlowQueryRetriever uses to read slow log data.
+type SlowQueryRetriever struct {
+	table       *model.TableInfo
+	outputCols  []*model.ColumnInfo
+	retrieved   bool
+	initialized bool
+	file        *os.File
+	fileLine    int
 }
 
-func dataForSlowLog(ctx sessionctx.Context) ([][]types.Datum, error) {
+func (e *SlowQueryRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.retrieved {
+		return nil, nil
+	}
+	if !e.initialized {
+		var err error
+		e.file, err = os.Open(sctx.GetSessionVars().SlowQueryFile)
+		if err != nil {
+			return nil, err
+		}
+		e.initialized = true
+	}
+	fullRows, err := e.dataForSlowLog(sctx)
```

Comment:
How about change `fullRows` to `rows` ，and change the `rows `in line 69 to retRows，because the `e.dataForSlowLog(sctx)` only return a part rows.

Reply:
great, done.